File: cpp_examples/async_example/yolov8_async_demo.cpp
// ============================================================
// yolov8_async_demo.cpp  --  YOLOv8 async producer/consumer demo
// Two threads: producer (read+letterbox+RunAsync) and the main
// thread (Wait+decode+draw+imshow). Geometry travels with each
// request so multiple frames can be in flight without stomping
// on each other. HighGUI stays on the main thread where the
// window was created. Queue depth 2 keeps the NPU fed without
// memory bloat. Trade-off: higher throughput, extra latency.
// Usage: ./yolov8_async_demo [--model path] [--source video|camera]
//                           [--path file] [--conf 0.3] [--iou 0.45]
// ============================================================

#include "config_loader.h"
#include "input_source.h"
#include "label_sets.h"
#include "ppu_decode.h"
#include "sdk_utils.h"
#include "visualizer.h"

#include <dxrt/dxrt_api.h>
#include <opencv2/highgui.hpp>

#include <atomic>
#include <chrono>
#include <condition_variable>
#include <cstdio>
#include <mutex>
#include <queue>
#include <thread>

namespace {

struct PendingRequest {
    int                 reqId;
    cv::Mat             frame;      // original BGR frame (owns pixels)
    cv::Mat             inputTensor;// keeps letterbox buffer alive until NPU done
    float               gain;
    cv::Point2f         pad;
    cv::Size            origSize;
};

class BoundedQueue {
public:
    explicit BoundedQueue(size_t cap) : m_cap(cap) {}

    bool push(PendingRequest&& item) {
        std::unique_lock<std::mutex> lk(m_mtx);
        m_notFull.wait(lk, [&]{ return m_q.size() < m_cap || m_stop; });
        if (m_stop) return false;
        m_q.push(std::move(item));
        m_notEmpty.notify_one();
        return true;
    }

    bool pop(PendingRequest& out) {
        std::unique_lock<std::mutex> lk(m_mtx);
        m_notEmpty.wait(lk, [&]{ return !m_q.empty() || m_done; });
        if (m_q.empty()) return false;
        out = std::move(m_q.front());
        m_q.pop();
        m_notFull.notify_one();
        return true;
    }

    void markDone()  { std::lock_guard<std::mutex> lk(m_mtx); m_done = true; m_notEmpty.notify_all(); }
    void requestStop(){ std::lock_guard<std::mutex> lk(m_mtx); m_stop = true; m_notFull.notify_all(); m_notEmpty.notify_all(); }

private:
    size_t                      m_cap;
    std::queue<PendingRequest>  m_q;
    std::mutex                  m_mtx;
    std::condition_variable     m_notFull, m_notEmpty;
    bool                        m_done = false;
    bool                        m_stop = false;
};

} // namespace

int main(int argc, char** argv) {
    auto p    = loadConfig("", argc, argv);
    auto labs = sdk::loadLabels(p.labelPath, labels::COCO80);

    int inputH = p.inputHeight, inputW = p.inputWidth;
    auto engine = sdk::loadEngine(p.modelPath, inputH, inputW);
    if (!engine) { std::fprintf(stderr, "[ERROR] Failed to load model\n"); return 1; }

    InputSource source(parseSourceType(p.sourceType), p.sourcePath, p.cameraIndex);
    if (!source.isOpened()) {
        std::fprintf(stderr, "[ERROR] Failed to open source: %s\n", p.sourcePath.c_str());
        return 1;
    }

    BoundedQueue queue(/*cap=*/2);
    std::atomic<bool> stopFlag{false};

    std::thread producer([&]() {
        cv::Mat frame;
        while (!stopFlag.load()) {
            if (!source.read(frame) || frame.empty()) break;
            PendingRequest r;
            r.frame    = frame.clone();
            auto lb    = sdk::letterbox(r.frame, inputH, inputW);
            r.inputTensor = lb.image;
            r.gain     = lb.gain;
            r.pad      = lb.pad;
            r.origSize = r.frame.size();
            r.reqId    = engine->RunAsync(r.inputTensor.data, nullptr, nullptr);
            if (!queue.push(std::move(r))) break;
        }
        queue.markDone();
    });

    std::fprintf(stderr, "[INFO] Async pipeline started. Press 'q' or ESC to quit.\n");
    cv::namedWindow(p.windowTitle, cv::WINDOW_NORMAL);

    auto tLast = std::chrono::steady_clock::now();
    double emaFps = 0.0;

    PendingRequest r;
    while (queue.pop(r)) {
        auto outs = engine->Wait(r.reqId);
        auto dets = ppu::decodeYolov8Ppu(outs, inputW, inputH, p.confThreshold, p.iouThreshold);

        std::vector<cv::Rect2f> boxes;
        std::vector<float>      scores;
        std::vector<int>        classIds;
        for (const auto& d : dets) boxes.push_back(d.box);
        auto boxesPx = sdk::unletterboxBoxes(boxes, r.gain, r.pad, r.origSize);
        for (size_t i = 0; i < dets.size(); ++i) {
            scores.push_back(dets[i].score);
            classIds.push_back(dets[i].classId);
        }
        vis::drawDetections(r.frame, boxesPx, scores, classIds, labs);

        auto tNow = std::chrono::steady_clock::now();
        double dt = std::chrono::duration<double>(tNow - tLast).count();
        tLast = tNow;
        if (dt > 0.0) emaFps = emaFps ? 0.9 * emaFps + 0.1 / dt : 1.0 / dt;
        if (p.showFps) vis::drawFps(r.frame, emaFps);

        cv::imshow(p.windowTitle, r.frame);
        int key = cv::waitKey(1) & 0xFF;
        if (key == 'q' || key == 27) { stopFlag.store(true); queue.requestStop(); break; }
    }

    stopFlag.store(true);
    queue.requestStop();
    if (producer.joinable()) producer.join();
    cv::destroyAllWindows();
    return 0;
}

File: python_examples/async_example/yolov8_async_demo.py
"""YOLOv8 Async Pipeline Demo — DeepX SDK
--------------------------------------------
Description: Same YOLOv8 model as the sync demo, but driven by a 3-stage
             pipeline that overlaps preprocess → inference → postprocess
             across worker threads. Each stage runs on its own thread and
             passes work via bounded queues. This dramatically increases
             throughput on long videos but adds a few frames of end-to-end
             latency, so it is the right tool for "process this batch as
             fast as possible" workloads and the wrong tool for low-latency
             interactive UIs.

             Trade-off summary:
                 sync mode    -> low latency,  low throughput
                 async mode   -> higher latency, higher throughput

Input      : webcam / rpicam / video file / image file
Output     : OpenCV window with bounding boxes; FPS counted at the renderer.

Usage:
    python yolov8_async_demo.py                              # uses config.py
    python yolov8_async_demo.py --source video --path drive.mp4
    python yolov8_async_demo.py --model yolov8.dxnn --queue 4
"""

from __future__ import annotations

import argparse
import os
import queue
import sys
import threading
import time
from typing import Tuple

import cv2
import numpy as np

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

import config
from utils import sdk
from utils.fps_counter import FPSCounter
from utils.input_source import InputSource
from utils.label_sets import COCO80
from utils.visualizer import draw_detections

DEFAULT_QUEUE_SIZE = 4   # How many frames may be in flight at once.
SENTINEL = object()      # Marker used to tell workers it is time to quit.


class YOLOv8Async:
    """YOLOv8 model wrapper that exposes both sync and async run paths.

    Most demos in this repo just call `engine.run([tensor])`. This one
    instead calls `engine.run_async(...)` to get a request id and later
    `engine.wait(req_id)` to fetch the outputs. That split is what lets
    a separate thread overlap host-side work with NPU work.
    """

    def __init__(self, model_path: str, conf_threshold: float,
                 iou_threshold: float) -> None:
        self.engine, self.input_h, self.input_w = sdk.load_engine(model_path)
        self.conf_threshold = conf_threshold
        self.iou_threshold = iou_threshold

    # The geometry context (gain, pad, src shape) lives on each work item
    # rather than on `self`, because in async mode multiple frames may be
    # in flight at once and they would otherwise stomp on each other.

    def preprocess(self, frame_bgr: np.ndarray):
        src_shape = frame_bgr.shape[:2]
        rgb = sdk.bgr_to_rgb(frame_bgr)
        padded, gain, pad = sdk.letterbox(rgb, (self.input_h, self.input_w))
        return padded, {"gain": gain, "pad": pad, "src_shape": src_shape}

    def postprocess(self, output_tensors, geom):
        outputs = np.transpose(np.squeeze(output_tensors[0]))
        cls_scores = outputs[:, 4:]
        scores = np.max(cls_scores, axis=1)
        class_ids = np.argmax(cls_scores, axis=1)
        boxes = sdk.cxcywh_to_xyxy(outputs[:, :4])

        keep = sdk.nms(boxes, scores, self.conf_threshold, self.iou_threshold)
        if keep.size == 0:
            empty = np.empty((0,), dtype=np.float32)
            return np.empty((0, 4), dtype=np.float32), empty, empty.astype(np.int64)

        boxes = sdk.unletterbox_boxes(
            boxes[keep], geom["gain"], geom["pad"], geom["src_shape"]
        )
        return boxes, scores[keep], class_ids[keep]


def _drain_queue(q: queue.Queue) -> None:
    """Pop everything from a queue without blocking. Used at shutdown."""
    try:
        while True:
            q.get_nowait()
    except queue.Empty:
        pass


def run_pipeline(model: YOLOv8Async, source: InputSource,
                 labels, queue_size: int, show: bool) -> None:
    """Three-stage pipeline: preprocess → wait → postprocess+draw.

    Each `queue.Queue(maxsize=queue_size)` acts as backpressure: when a
    downstream stage is slow, upstream stages naturally block on `put()`
    instead of growing memory without bound.
    """

    pre_q: "queue.Queue[Tuple[np.ndarray, np.ndarray, dict]]" = queue.Queue(
        maxsize=queue_size
    )
    wait_q: "queue.Queue[Tuple[np.ndarray, object, dict]]" = queue.Queue(
        maxsize=queue_size
    )

    stop = threading.Event()
    fps = FPSCounter()

    def producer():
        """Read frames + run preprocess, then submit to the NPU."""
        try:
            while not stop.is_set():
                ok, frame = source.read()
                if not ok or frame is None:
                    break
                input_tensor, geom = model.preprocess(frame)
                # Submit to the NPU before blocking on the queue so that
                # the device starts working as early as possible.
                req_id = model.engine.run_async([input_tensor])
                pre_q.put((frame, input_tensor, geom, req_id))
        except Exception as exc:
            print(f"[ERROR] Producer thread: {exc}")
            stop.set()
        finally:
            pre_q.put(SENTINEL)

    def waiter():
        """Block on each request id and forward the raw outputs."""
        try:
            while True:
                item = pre_q.get()
                if item is SENTINEL or stop.is_set():
                    break
                frame, input_tensor, geom, req_id = item
                # input_tensor is kept alive in the tuple so its memory
                # cannot be reused by NumPy while the NPU is still reading.
                outputs = model.engine.wait(req_id)
                wait_q.put((frame, outputs, geom))
        except Exception as exc:
            print(f"[ERROR] Waiter thread: {exc}")
            stop.set()
        finally:
            wait_q.put(SENTINEL)

    # The consumer (postprocess + draw + display) runs on the main thread
    # because OpenCV's HighGUI must be called from the thread that created
    # the window. Workers above only touch numpy and SDK calls.
    print("[INFO] Async pipeline started. Press 'q' or ESC to quit.")
    threads = [
        threading.Thread(target=producer, daemon=True),
        threading.Thread(target=waiter, daemon=True),
    ]
    for t in threads:
        t.start()

    try:
        while True:
            item = wait_q.get()
            if item is SENTINEL:
                print("[INFO] Input source exhausted.")
                break
            frame, outputs, geom = item

            try:
                boxes, scores, class_ids = model.postprocess(outputs, geom)
            except Exception as exc:
                print(f"[WARN] Postprocess failed for one frame: {exc}")
                continue

            draw_detections(frame, boxes, scores, class_ids, labels, config)

            fps.update()
            if config.SHOW_FPS:
                fps.draw(frame)

            if show:
                cv2.imshow(config.WINDOW_NAME, frame)
                key = cv2.waitKey(1) & 0xFF
                if key in (ord("q"), 27):
                    print("[INFO] Quit requested by user.")
                    break

    except KeyboardInterrupt:
        print("[INFO] Interrupted (Ctrl+C).")
    finally:
        stop.set()
        # Wake any worker blocked on a full queue.
        _drain_queue(pre_q)
        _drain_queue(wait_q)
        for t in threads:
            t.join(timeout=2.0)
        source.release()
        cv2.destroyAllWindows()


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="YOLOv8 Async Demo — DeepX SDK")
    p.add_argument("--source", type=str, default=config.INPUT_SOURCE,
                   choices=list(InputSource.SUPPORTED))
    p.add_argument("--path", type=str, default=None)
    p.add_argument("--model", type=str, default=config.MODEL_PATH)
    p.add_argument("--labels", type=str, default=config.LABEL_PATH)
    p.add_argument("--conf", type=float, default=config.CONFIDENCE_THRESHOLD)
    p.add_argument("--iou", type=float, default=config.IOU_THRESHOLD)
    p.add_argument("--queue", type=int, default=DEFAULT_QUEUE_SIZE,
                   help="Max in-flight frames between pipeline stages.")
    p.add_argument("--no-display", dest="display", action="store_false")
    p.set_defaults(display=True)
    return p.parse_args()


def _resolve_path(args: argparse.Namespace) -> str:
    if args.path is not None:
        return args.path
    if args.source == "video":
        return config.VIDEO_PATH
    if args.source == "image":
        return config.IMAGE_PATH
    return ""


def main() -> None:
    args = parse_args()
    labels = sdk.load_labels(args.labels, COCO80)
    model = YOLOv8Async(args.model, args.conf, args.iou)
    source = InputSource(args.source, path=_resolve_path(args),
                         webcam_index=config.WEBCAM_INDEX)

    run_pipeline(model, source, labels, queue_size=max(1, args.queue),
                 show=args.display)


if __name__ == "__main__":
    main()
